### PR TITLE
Leader eviction

### PR DESCRIPTION
**Pull request by jbsv**

### CODE REVIEW

**Review by nkcr — COMMENTED** [review]
Looks good overall, the logic is right !

You made an update on the transaction implementation, but it would rather not touch this module. IMO it is entirely the responsibility of the pool to keep stats about transaction, and not the transaction module.

In that sense I would keep the transaction module untouched, and make the pool module wrap transactions and keep its own information. Something like `statsTrasnaction`:

```
type transactions []statTransaction

type statTransaction struct {
   txn.Transaction

   stats stats
}

type stats struct {
   inclusionTime time.Time
}
```

File: core/txn/pool/gatherer.go [hunk 1]
```diff
@@ -214,6 +158,35 @@ func (g *simpleGatherer) Wait(ctx context.Context, cfg Config) []txn.Transaction
 	}
 }
 
+func (g *simpleGatherer) Stats() Stats {
```
Comment (jbsv) [inline core/txn/pool/gatherer.go:161]: function docs

File: core/txn/pool/gatherer.go [hunk 2]
```diff
@@ -214,6 +158,35 @@ func (g *simpleGatherer) Wait(ctx context.Context, cfg Config) []txn.Transaction
...
+	for _, tx := range txs {
+		if tx.insertionTime.Before(stats.OldestTx) {
+			stats.OldestTx = tx.insertionTime
+		}
+	}
+
+	return stats
+}
+
+func (g *simpleGatherer) ResetStats() {
```
Comment (jbsv) [inline core/txn/pool/gatherer.go:180]: function docs

File: core/txn/pool/gossip/mod.go [hunk 3]
```diff
@@ -94,6 +87,14 @@ func (p *Pool) Gather(ctx context.Context, cfg pool.Config) []txn.Transaction {
 	return p.gatherer.Wait(ctx, cfg)
 }
 
+func (p *Pool) Stats() pool.Stats {
```
Comment (jbsv) [inline core/txn/pool/gossip/mod.go:90]: function docs

File: core/txn/pool/gossip/mod.go [hunk 4]
```diff
@@ -94,6 +87,14 @@ func (p *Pool) Gather(ctx context.Context, cfg pool.Config) []txn.Transaction {
...
+	return p.gatherer.Stats()
+}
+
+func (p *Pool) ResetStats() {
```
Comment (jbsv) [inline core/txn/pool/gossip/mod.go:94]: function docs

File: core/txn/pool/mem/mod.go [hunk 5]
```diff
@@ -59,19 +52,27 @@ func (s *Pool) Remove(tx txn.Transaction) error {
...
 // Close implements pool.Pool. It cleans the resources of the gatherer.
-func (s *Pool) Close() error {
-	s.gatherer.Close()
+func (p *Pool) Close() error {
+	p.gatherer.Close()
 
 	return nil
 }
+
+func (p *Pool) Stats() pool.Stats {
```
Comment (jbsv) [inline core/txn/pool/mem/mod.go:72]: function docs

File: core/txn/pool/mem/mod.go [hunk 6]
```diff
@@ -59,19 +52,27 @@ func (s *Pool) Remove(tx txn.Transaction) error {
...
+	return p.gatherer.Stats()
+}
+
+func (p *Pool) ResetStats() {
```
Comment (jbsv) [inline core/txn/pool/mem/mod.go:76]: function docs

File: core/txn/pool/gatherer.go [hunk 7]
```diff
@@ -86,6 +33,9 @@ type Gatherer interface {
 
 	// Close closes current operations and cleans the resources.
 	Close()
+
+	Stats() Stats
```
Comment (jbsv) [inline core/txn/pool/gatherer.go:37]: docs

File: core/txn/pool/gatherer.go [hunk 8]
```diff
@@ -1,75 +1,22 @@
...
 	"context"
-	"sort"
-	"sync"
-
 	"go.dedis.ch/dela/core/access"
 	"go.dedis.ch/dela/core/txn"
 	"go.dedis.ch/dela/core/validation"
 	"golang.org/x/xerrors"
+	"sync"
+	"time"
```
Comment (pierluca) [inline core/txn/pool/gatherer.go:10]: the current ordering of the imports seems a bit off.
could it be that you're not using goimports ?

File: core/txn/pool/stats.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+// transactionStats enhances a transaction with some statistics
+// to allow the detection of rotten transactions in a pool.
+type transactionStats struct {
+	txn.Transaction
+	insertionTime time.Time
+}
+
+// ResetStats resets the insertion time to now.
+// It is used when a leader view change is necessary.
```
Comment (pierluca) [inline core/txn/pool/stats.go:16]: This looks inaccurate. How about 
```suggestion
// It is used when a leader view change is initiated.
```
?

File: core/txn/signed/mod.go [hunk 10]
```diff
@@ -21,6 +17,8 @@ import (
 	"go.dedis.ch/dela/serde"
 	"go.dedis.ch/dela/serde/registry"
 	"golang.org/x/xerrors"
+	"io"
+	"sort"
```
Comment (pierluca) [inline core/txn/signed/mod.go:21]: this again looks like you're not using goimports

File: go.mod [hunk 11]
```diff
@@ -15,10 +15,10 @@ require (
 	github.com/urfave/cli/v2 v2.2.0
```
Comment (pierluca) [inline go.mod:15]: why upgrade the dependencies with this PR ? 🤔

File: core/txn/pool/gatherer_test.go [hunk 12]
```diff
@@ -2,25 +2,24 @@ package pool
...
-	"sync"
-	"testing"
-
 	"github.com/stretchr/testify/require"
 	"go.dedis.ch/dela/core/access"
 	"go.dedis.ch/dela/core/txn"
 	"go.dedis.ch/dela/core/validation"
 	"go.dedis.ch/dela/internal/testing/fake"
+	"sync"
+	"testing"
```
Comment (pierluca) [inline core/txn/pool/gatherer_test.go:11]: goimports :)

File: core/ordering/cosipbft/mod.go [hunk 13]
```diff
@@ -223,9 +231,9 @@ func NewService(param ServiceParam, opts ...ServiceOption) (*Service, error) {
 		actor:                    actor,
 		val:                      param.Validation,
 		verifierFac:              param.Cosi.GetVerifierFactory(),
-		timeoutRound:             RoundTimeout,
-		timeoutRoundAfterFailure: RoundTimeout,
-		timeoutViewchange:        RoundTimeout,
+		timeoutRound:             DefaultRoundTimeout,
+		timeoutRoundAfterFailure: DefaultRoundTimeoutAfterFailure,
+		timeoutViewchange:        DefaultTimeoutBeforeViewchange,
```
Comment (pierluca) [inline core/ordering/cosipbft/mod.go:236]: do we really want to change these values in production?
It seems like a rather impactful change.

File: core/ordering/cosipbft/mod_test.go [hunk 14]
```diff
@@ -122,10 +114,74 @@ func TestService_Scenario_ViewChange(t *testing.T) {
...
+	defer clean()
+
+	signer := nodes[0].signer
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	initial := ro.Take(mino.RangeFilter(0, 4)).(crypto.CollectiveAuthority)
+
+	err := nodes[0].service.Setup(ctx, initial)
```
Comment (pierluca) [inline core/ordering/cosipbft/mod_test.go:162]: why can't this just be 
```suggestion
	err := nodes[0].service.Setup(ctx, ro)
```
?

File: core/ordering/cosipbft/mod_test.go [hunk 15]
```diff
@@ -122,10 +114,74 @@ func TestService_Scenario_ViewChange(t *testing.T) {
...
+	require.NoError(t, err)
+
+	leader, err := nodes[0].service.pbftsm.GetLeader()
+	require.NoError(t, err)
+	require.Equal(t, leader, nodes[0].onet.GetAddress())
+
+	err = nodes[0].pool.Add(makeTx(t, 0, signer))
+	require.NoError(t, err)
+
+	time.Sleep(DefaultRoundTimeout + 1*time.Second)
```
Comment (pierluca) [inline core/ordering/cosipbft/mod_test.go:172]: I'm not sure that `DefaultRoundTimeout` is the correct time-out now, given the split in 3 different time-outs and the viewchange one being longer. Is it ?

File: core/ordering/cosipbft/mod.go [hunk 16]
```diff
@@ -465,109 +476,128 @@ func (s *Service) doRound(ctx context.Context) error {
...
 	leader, err := s.pbftsm.GetLeader()
 	if err != nil {
 		return xerrors.Errorf("reading leader: %v", err)
 	}
 
-	timeout := s.timeoutRound
-	if s.failedRound {
-		timeout = s.timeoutRoundAfterFailure
+	if s.me.Equal(leader) {
+		s.logger.Info().Msgf("doLeaderRound with timeout %f", float32(timeout)/1000000000)
```
Comment (pierluca) [inline core/ordering/cosipbft/mod.go:490]: why go through such a complicated expression ? 
Also, shouldn't this be a debug message ? 
```suggestion
		s.logger.Debug().Msgf("doLeaderRound with timeout %s", timeout)
```

Maybe the message could be improved too, as this is not very human friendly.

File: core/ordering/cosipbft/mod.go [hunk 17]
```diff
@@ -465,109 +476,128 @@ func (s *Service) doRound(ctx context.Context) error {
...
+		return s.doLeaderRound(ctx, roster, timeout)
 	}
 
-	for !s.me.Equal(leader) {
-		// Only enters the loop if the node is not the leader. It has to wait
-		// for the new block, or the round timeout, to proceed.
+	s.logger.Info().Msgf("doFollowerRound with timeout %f", float32(timeout)/1000000000)
```
Comment (pierluca) [inline core/ordering/cosipbft/mod.go:494]: same

File: core/ordering/cosipbft/mod.go [hunk 18]
```diff
@@ -465,109 +476,128 @@ func (s *Service) doRound(ctx context.Context) error {
...
 		}
-	}
 
-	ctx, cancel := context.WithTimeout(ctx, timeout)
-	defer cancel()
+		s.logger.Info().Msgf("view change for %d", viewMsg.GetLeader())
 
-	s.logger.Debug().Uint64("index", s.blocks.Len()).Msg("round has started")
+		return nil
+	case <-s.events:
```
Comment (pierluca) [inline core/ordering/cosipbft/mod.go:574]: It seems like readability / vertical spacing could be improved a little

```suggestion
		return nil
		
	case <-s.events:
```

File: core/ordering/cosipbft/mod.go [hunk 19]
```diff
@@ -465,109 +476,128 @@ func (s *Service) doRound(ctx context.Context) error {
...
+		s.failedRound = false
 
-	// Send a synchronization to the roster so that they can learn about the
-	// latest block of the chain.
-	err = s.sync.Sync(ctx, roster, blocksync.Config{MinHard: threshold.ByzantineThreshold(roster.Len())})
-	if err != nil {
-		return xerrors.Errorf("sync failed: %v", err)
+		// A block has been created meaning that the round is over.
+		return nil
+	case <-s.closing:
```
Comment (pierluca) [inline core/ordering/cosipbft/mod.go:581]: same

File: go.mod (hunk 11 above)
Comment (jbsv) [reply go.mod:15]: it's always done with go mod tidy

Comment (jbsv) [reply go.mod:15]: go mod tidy is always called with `make check

**Review by pierluca — COMMENTED** [review]
LGTM, just minor naming concern.

File: core/ordering/cosipbft/mod.go [hunk 20]
```diff
@@ -104,7 +106,7 @@ type Service struct {
 
 	timeoutRound             time.Duration
 	timeoutRoundAfterFailure time.Duration
-	timeoutViewchange        time.Duration
+	transactionTimeout       time.Duration
```
[the remark this answers is not in the document]
Comment (pierluca) [reply core/ordering/cosipbft/mod.go:109]: actually, re-reading this, the new name is better 😅

**Review by pierluca — APPROVED** [review]

**Pull request merged**